From 21fba649229b7599c4b42d3cbcf61d2ab96241cc Mon Sep 17 00:00:00 2001
From: Jacob Young <jacobly0@users.noreply.github.com>
Date: Wed, 10 Sep 2025 03:10:46 -0400
Subject: [PATCH 110/133] x86_64: fix safety crashes in `storeRegs`
X-Developer-Signature: v=1; a=openpgp-sha256; l=2491; i=xtex@astrafall.org;
 h=from:subject; bh=syeP2Za7edJ0S7V/BGKDmcWDe7A4l7LbBFSpZH9kPUU=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78nga3zy4asGvLi/+5PBZ39yUIymTD01c/oZ/3lTt/
 /NfRulod0xkYRDjYrAUU2QpMmzwZtVJ5xddVi4LM4eVCWSItEgDAwMDAwsDX25iXqmRjpGeqbah
 nqGhjpGOEQMXpwBMtfRsRoYbewI9lTava5229MJ7GWkf3o/n9UXPBogKux4rWmk0zeQgwz/l3Ot
 +T958uZd5K00oV/+IiETJsX7DrP0layNe7s3rkGAEAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

(cherry picked from commit cbbb67c5df46bc97f70f925772ce03eb11db5a1b)
---
 src/arch/x86_64/CodeGen.zig | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/src/arch/x86_64/CodeGen.zig b/src/arch/x86_64/CodeGen.zig
index c41f2e6b4a85..40dbe016db34 100644
--- a/src/arch/x86_64/CodeGen.zig
+++ b/src/arch/x86_64/CodeGen.zig
@@ -187009,7 +187009,13 @@ const Temp = struct {
                 },
                 .struct_type => {
                     assert(src_regs.len - part_index == std.math.divCeil(u32, src_abi_size, 8) catch unreachable);
-                    break :part_ty .u64;
+                    break :part_ty switch (src_abi_size) {
+                        0, 3, 5...7 => unreachable,
+                        1 => .u8,
+                        2 => .u16,
+                        4 => .u32,
+                        else => .u64,
+                    };
                 },
                 .tuple_type => |tuple_type| {
                     assert(tuple_type.types.len == src_regs.len);
@@ -187018,10 +187024,6 @@ const Temp = struct {
             };
             const part_size: u31 = @intCast(part_ty.abiSize(zcu));
             const src_rc = src_reg.class();
-            const part_bit_size = switch (src_rc) {
-                else => 8 * part_size,
-                .x87 => part_ty.bitSize(zcu),
-            };
             if (src_rc == .x87 or std.math.isPowerOfTwo(part_size)) {
                 // hack around linker relocation bugs
                 switch (ptr.tracking(cg).short) {
@@ -187030,7 +187032,15 @@ const Temp = struct {
                 }
                 const strat = try cg.moveStrategy(part_ty, src_rc, false);
                 try strat.write(cg, try ptr.tracking(cg).short.deref().mem(cg, .{
-                    .size = .fromBitSize(part_bit_size),
+                    .size = switch (src_rc) {
+                        else => .fromBitSize(8 * part_size),
+                        .x87 => switch (abi.classifySystemV(src_ty, zcu, cg.target, .other)[part_index]) {
+                            else => unreachable,
+                            .float => .dword,
+                            .float_combine, .sse => .qword,
+                            .x87 => .tbyte,
+                        },
+                    },
                     .disp = part_disp,
                 }), registerAlias(src_reg, part_size));
             } else {
-- 
2.52.0

